### PR TITLE
Add Schnorr signature

### PR DESCRIPTION
**Pull request by armatusmiles**

### CODE REVIEW

File: crypto/schnorr.c [hunk 1]
```diff
@@ -0,0 +1,78 @@
+#include "schnorr.h"
```
Comment (prusnak) [inline crypto/schnorr.c]: Add copyright header (the same as in `schnorr.h`)

File: crypto/schnorr.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  point_add(curve, &sG, &Q);
+
+  /* r = H(Q, kpub, m) */
+  bignum256 r;
+  calc_r(&Q, pub_key, msg, msg_len, &r);
+
+  if (bn_is_equal(&r, &sign->r)) return 0;  // success
+
+  return 10;
+}
```
Comment (prusnak) [inline crypto/schnorr.c]: Add newline at the end of the file.

File: crypto/schnorr.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+// sign/verify returns 0 if operation succeeded
+
+// k is a random from [1, ..., order-1]
+int schnorr_sign(const ecdsa_curve *curve, const uint8_t *priv_key,
+                 const bignum256 *k, const uint8_t *msg, const uint32_t msg_len,
+                 schnorr_sign_pair *result);
+int schnorr_verify(const ecdsa_curve *curve, const uint8_t *pub_key,
+                   const uint8_t *msg, const uint32_t msg_len,
+                   const schnorr_sign_pair *sign);
+#endif
```
Comment (prusnak) [inline crypto/schnorr.h]: Add newline at the end of the file.

File: crypto/tests/test_check.c [hunk 4]
```diff
@@ -8347,6 +8348,191 @@ START_TEST(test_compress_coords) {
...
 END_TEST
 
+START_TEST(test_schnorr_sign_verify) {
+  static struct {
+    const char *message;
+    const char *priv_key;
+    const char *k_hex;
+    const char *s_hex;
+    const char *r_hex;
+  } test_cases[] = {
```
Comment (prusnak) [inline crypto/tests/test_check.c]: Where do these test vectors come from?

If these are your own, can you please also add some that are not your own? Simple google search for "schnorr test vectors" returns some results such as:

* https://raw.githubusercontent.com/sipa/bips/bip-schnorr/bip-schnorr/test-vectors.csv
* https://gist.github.com/kallewoof/5d623445802a84f17cc7ff5572109074

File: crypto/schnorr.c [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  uint8_t digest[SHA256_DIGEST_LENGTH];
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, Q_compress, 33);
+  sha256_Update(&ctx, pub_key, 33);
+  sha256_Update(&ctx, msg, msg_len);
+  sha256_Final(&ctx, digest);
+  bn_read_be(digest, r);
+}
+
+// returns 0 if verification succeeded
```
Comment (prusnak) [inline crypto/schnorr.c]: This should read `returns 0 if signing succeeded`

File: crypto/schnorr.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  while (bn_is_less(&curve->order, &result->s) ||
+         bn_is_equal(&curve->order, &result->s)) {
+    bn_mod(&result->s, &curve->order);
+  }
+
+  return 0;
+}
+
+int schnorr_verify(const ecdsa_curve *curve, const uint8_t *pub_key,
```
Comment (prusnak) [inline crypto/schnorr.c]: Maybe add comment `returns 0 if verification succeeded`

File: crypto/schnorr.c (hunk 1 above)
Comment (armatusmiles) [reply crypto/schnorr.c]: done

Comment (armatusmiles) [reply crypto/schnorr.c]: > Add copyright header (the same as in `schnorr.h`)

done

File: crypto/schnorr.c (hunk 6 above)
Comment (armatusmiles) [reply crypto/schnorr.c]: > Please add comment `returns 0 if verification succeeded`

done

File: crypto/tests/test_check.c (hunk 4 above)
Comment (armatusmiles) [reply crypto/tests/test_check.c]: I got test vectors from results [this](https://github.com/Zilliqa/Zilliqa/blob/master/src/libCrypto/Schnorr.cpp#L88)  function.
By your first reference they don't use a random K.
By second reference - message in hex format - uncomfortably for me. Need rewrite tests for use it

**Comment by prusnak** [comment]
Squashed and merged in https://github.com/trezor/trezor-firmware/commit/8114e0f9fc7af5d41ac73d69209b4065d3638327

Thanks!

**Comment by prusnak** [comment]
Hey @armatusmiles

I have reverted the commit and removed the Schnorr implementation from master. @onvej-sl found out the Schnorr implementation you provided is insecure and will tell you more.

**Comment by armatusmiles** [comment]
@prusnak Hello. Thank you for feedback. I’m sorry.

**Comment by hewigovens** [comment]
Hi @prusnak @onvej-sl 

could you please share more details about the implementation? I believe @armatusmiles implemented Zilliqa scheme (https://github.com/Zilliqa/Zilliqa/blob/master/src/libCrypto/Schnorr.cpp), which is audited by NCC group (Zilliqa team tells us)

Thanks

**Comment by onvej-sl** [comment]
Hi @hewigovens,

I found the implementation insecure for the following reasons:
1. In `schnorr_verify`, you should check that both r and s parts of the signature are **less** than the curve order. That's how Schnorr signature is usually implemented, Zilliqa not being an exception.
By the way, `scalar_multiply` **assumes** that exponent is less than the order, so in a special case verification failed with an assertion error.
2. You represent a signature by two `bignum256` numbers. Since every number has several `bignum256` representation (which may be partly reduced, reduced or normalised), great care must be taken when dealing with it. The raw format (an output of `bn_write_be`) would be better.
Assume `bignum256 r` is the r part of a signature on `secp256k1` curve with `r.val = {0x10364142, 0x3f497a33, 0x348a03bb, 0x2bb739ab, 0x3ffffeba, 0x3fffffff, 0x3fffffff, 0xffffffff, 0x0000fffc}`. Such `r` pass through the condition `if (bn_is_less(&curve->order, &sign->r)) return 4;`, however it shouldn't.
3. Your nonce is user defined. Since a nonce reuse reveals the private key, it should be generated randomly or deterministically (for example as defined in RFC 6979) right in `ecdsa_sign`.

**Comment by hewigovens** [comment]
@onvej-sl Thanks for your elaborations, for `3` we did use `generate_k_rfc6979` to generate the nonce in a wrapper method,  it's just not ready to push to upstream. we'll address `1` and `2` shortly

**Comment by ansnunez** [comment]
Hi @onvej-sl for `1` I think the checks are already there?  The 4 lines starting from https://github.com/TrustWallet/wallet-core/blob/1d67cd877ae495a603715a2b649c1271cbc8ff83/trezor-crypto/src/schnorr.c#L116

In Zilliqa, though, we also check for `r` and `s` being negative: https://github.com/Zilliqa/Zilliqa/blob/4ed7a542efc98ef00ae38688f66833492786536e/src/libCrypto/Schnorr.cpp#L295

**Comment by ansnunez** [comment]
@onvej-sl for `2`, I think `calc_r` is missing a call to `bn_nnmod` to avoid the problem in your example.

Here is the equivalent code in Zilliqa:
https://github.com/Zilliqa/Zilliqa/blob/4ed7a542efc98ef00ae38688f66833492786536e/src/libCrypto/Schnorr.cpp#L158

If we put this code in place, would it then be OK to keep the two `bignum256` representation?

**Comment by onvej-sl** [comment]
> Hi @onvej-sl for `1` I think the checks are already there? The 4 lines starting from https://github.com/TrustWallet/wallet-core/blob/1d67cd877ae495a603715a2b649c1271cbc8ff83/trezor-crypto/src/schnorr.c#L116

It is not. What if `r` or `s` is equal to the order?

> @onvej-sl for 2, I think calc_r is missing a call to bn_nnmod to avoid the problem in your example.

It won't solve the problem. The problem is that the `r` I provided is not normalized. Please see `bignum.c` comments for normalized, reduced and partly reduced numbers.

> If we put this code in place, would it then be OK to keep the two bignum256 representation?

What is the reason you prefer `bignum256` over raw format? I think I've proven it not easy to work with `bignum256`.

**Comment by hewigovens** [comment]
@onvej-sl could you please review this [pr](https://github.com/TrustWallet/wallet-core/pull/476 )? @ansnunez tries to fix all the concerns.

Thanks!

**Comment by onvej-sl** [comment]
> @onvej-sl could you please review this [pr](https://github.com/TrustWallet/wallet-core/pull/476)? @ansnunez tries to fix all the concerns.

Good job! It seems to be OK. Are you going to push it to trezor-firmware?

**Comment by hewigovens** [comment]
Thanks, we will push it to upstream later

**Pull request closed without merging**